rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll

```diff
@@ -84,7 +84,10 @@ module Impl {
     // an enum constant (e.g. `None`). This excludes static and constant variables (UPPERCASE),
     // which we don't appear to recognize yet anyway. This also assumes programmers follow the
     // naming guidelines, which they generally do, but they're not enforced.
-    not name.charAt(0).isUppercase()
+    not name.charAt(0).isUppercase() and
+    // exclude parameters from functions without a bodies as these are trait method declarations
+    // without implementations
+    not exists(Function f | not f.hasBody() and f.getParamList().getAParam().getPat() = p)
```

Comment:
I'm not sure this is entirely correct. What about tuple parameters. For example:
```rust
fn foo(point@(x, y): (i32, i32)) { ... }
```

I'm also not entirely sure we don't want those parameters to be variables. They might be useful for dataflow when binding arguments to parameters in a function call.

Reply:
I'd say we bring this in as is and leave this for a follow up though. I think we can already a good chunk of FPs with this.

Reply:
ah, wait, thinking more clearly about it, yeah, maybe it's better not to put this restriction at the `VariableImpl` level, but something more specific for the unused variable/value query implementations.

Reply:
@aibaars See the comment I wrote here: https://github.com/github/codeql/pull/17847#discussion_r1816762685. This is only done for functions without bodies and patterns are not allowed for parameters in such functions. Rust will error for your example if you remove the body and put it in a trait.

> I'm also not entirely sure we don't want those parameters to be variables. They might be useful for dataflow when binding arguments to parameters in a function call.

Also for functions that have no body? If that is the case we should do the similar thing but directly in the two queries.

Reply:
Ah, didn't spot that comment, in that case it should be fine.

Whether or not it is useful to consider the parameters of trait functions depends a bit on how we connect the arguments to parameters. If we first connect the arguments of a call to the parameters of the function defined in a trait and subsequently link them to all suitable `impl`s of the function then we might end up wondering why nothing works ;-)  I guess if you directly link the arguments of a call to all suitable `impl`s then things would just work. Just keep this in mind when implementing dataflow calls.